_drafts/dealing-with-large-symbol-files.md

```diff
@@ -0,0 +1,592 @@
+program (using the excellent [hyperfine](https://github.com/sharkdp/hyperfine)
+benching tool):
+
+```bash
+❯ hyperfine 'gdb -batch -ex quit hello.debug' 'gdb -batch -ex quit hello.debug.compressed'
+Benchmark 1: gdb -batch -ex quit hello.debug
+  Time (mean ± σ):     418.8 ms ±   9.3 ms    [User: 350.2 ms, System: 67.6 ms]
+  Range (min … max):   399.8 ms … 428.9 ms    10 runs
+
+Benchmark 2: gdb -batch -ex quit hello.debug.compressed
+  Time (mean ± σ):     425.9 ms ±  14.7 ms    [User: 365.6 ms, System: 60.0 ms]
+  Range (min … max):   407.7 ms … 451.3 ms    10 runs
+
+Summary
+  'gdb -batch -ex quit hello.debug' ran
+    1.02 ± 0.04 times faster than 'gdb -batch -ex quit hello.debug.compressed'
+```
+
+With larger files it might be more significant, but I haven't tested that.
+
+Note that it's possible to _decompress_ the debug sections with `objcopy` using
+the `--decompress-debug-sections` flag.
+
+## Compress Debug Info with `dwz`
+
+The `dwz` tool available [here](https://sourceware.org/dwz/) (and from many
+package managers) performs some nice optimizations of DWARF info (the
+fundamental format of the debug info data). You can read some more details in
+the man page, either by downloading the source and loading it:
+
+```bash
+❯ man -l dwz.1
+```
+
+Or from one of the many online copies, for example
+[Debian Linux](https://manpages.debian.org/testing/dwz/dwz.1.en.html).
+
+Running it on our test binary:
+
+```bash
+❯ dwz hello.debug -o hello.debug.dwz
+❯ ls -l hello.debug*
+-rwxrwxr-x 1 noah noah 42336 Mar  2 12:30 hello.debug
+-rwxrwxr-x 1 noah noah 26144 Mar  2 14:11 hello.debug.compressed
+-rwxrwxr-x 1 noah noah 42304 Mar  2 14:22 hello.debug.dwz
+-rwxrwxr-x 1 noah noah 14568 Mar  2 14:07 hello.debug.stripped
+```
+
+Ok, that didn't really improve much 😕 it's possible the debug info for such a
+small program is already pretty optimized, or maybe `dwz` doesn't bother
+optimizing below a threshold (I haven't checked).
+
+Let's try it again on a large binary, the `cmake` program compiled with
+`CMAKE_BUILD_TYPE=Debug`:
+
+```bash
+# use dwz to optimize the debug information in 'bin/cmake'
+❯ dwz bin/cmake -o cmake.dwz
+# use objcopy to compress debug info in the result
+❯ objcopy --compress-debug-sections cmake.dwz cmake.dwz.compressed
+# and for completeness, also compress debug info in the original cmake file
+❯ objcopy --compress-debug-sections bin/cmake cmake.compressed
+# let's take a look at file sizes now
+❯ ls -lh bin/cmake cmake.dwz cmake.dwz.compressed cmake.compressed
+-rwxrwxr-x 1 noah noah 179M Mar  2 14:46 bin/cmake
+-rwxrwxr-x 1 noah noah  88M Mar  2 15:45 cmake.compressed
+-rwxrwxr-x 1 noah noah 129M Mar  2 14:47 cmake.dwz
+-rwxrwxr-x 1 noah noah  70M Mar  2 14:48 cmake.dwz.compressed
+```
+
+To summarize:
+
+- `dwz` by itself is able to save about 50MB from the original file (72% of the
+  original)
+- `--compress-debug-sections` alone does a pretty good job, saving 91MB (down to
```

Comment:
I found this a little bit confusing: it reads as if `dwz` saved 50MB and `--compress-debug-sections` saved 91MB (so the latter is way better), but it also says that `dwz` saved 72% of the original, and `compress-debug-sections` saved 49% (so the latter is worse?). I think what you mean is that the resulting binary is 72%/49% of the original, but it's not super clear.

Reply:
Good call, this section is pretty awkward, might make better sense as a table